### ISSUE TITLE
Transparent Borders on Windows 11

### ISSUE THREAD

**Issue by creikey** [slynch8/10x#907]
Looks like this was closed by https://github.com/slynch8/10x/issues/632 but I still have the bug on the latest version 0.32.19: 

![image](https://user-images.githubusercontent.com/28424206/169485983-d187465f-7e27-4b25-aaa0-94ac47d08309.png)

**Comment by slynch8** [comment]
I still can't reproduce this, can I send you some builds to try?

**Comment by slynch8** [comment]
can you see if this makes a difference
https://e1.pcloud.link/publink/show?code=XZhqVRZlUR9yapcQqkVGyqrw5HLo7MrvfyX

**Comment by slynch8** [comment]
here are another couple of builds. I've made the dropshadow red to make it easier to see what is going on
https://e1.pcloud.link/publink/show?code=XZgNVRZvhQky8qF0Dk0mO1TWTXBPRNtKgaX

https://e1.pcloud.link/publink/show?code=XZCNVRZvuGTsVmb30JjOi3dmrDgx8Ljaidy

**Comment by slynch8** [comment]
try this one first
https://e1.pcloud.link/publink/show?code=XZiaVRZopvtFIauajVW9DdzAsmtxyEf3q8X

**Comment by slynch8** [comment]
one more
https://e1.pcloud.link/publink/show?code=XZIrVRZgy2eMc1gNBHJXmlUVBmeY0itSA0V

**Comment by creikey** [comment]
> 

![image](https://user-images.githubusercontent.com/28424206/169527057-f5ba9938-2289-42d2-a3fe-30711d5dfe32.png)
border fix 5 produces this

**Comment by creikey** [comment]
border fix 4 behaves the same as 5

**Comment by creikey** [comment]
![image](https://user-images.githubusercontent.com/28424206/169527549-96d47cc6-0b99-45b0-b1bd-44771f292fa2.png)

border fix 1 has a blue outline like this that disappears when maximized, but it still has the border when maximized, like this:
![image](https://user-images.githubusercontent.com/28424206/169527614-b4fc6be7-610e-4313-8fc7-a9ba9c1e5335.png)

**Comment by creikey** [comment]
![image](https://user-images.githubusercontent.com/28424206/169528301-fd351b89-9aac-49f4-bfb8-625cb1d87b5b.png)

Border fix 2 looks like this, maximized the shadow goes away but the border remains

**Comment by creikey** [comment]
Border fix 3 behaves the same as border fix 5

**Comment by creikey** [comment]
@slynch8 as you can't reproduce on your machine would you like to remote into mine with teamviewer or something and try to fix the source code being able to see your changes instantly?

**Comment by slynch8** [comment]
Thanks for trying all those. I think I have have found it
https://e1.pcloud.link/publink/show?code=XZd50RZGNK66wjgsWfSTUhPgbs8vLPJPj7X

If that doesn't work remoting in would be super helpful thank you. I've not used teamviewer before but I'm willing to give it a try.

**Comment by BuzzwordChief** [comment]
I'am also experiencing this bug. Weirdly not on my old win11 laptop.

Checked to see if it was because of the iGPU (Intel Xe) but switching to my dedicated GPU (3050) yielded the same result.

**Comment by slynch8** [comment]
can you give this build a try:
https://e1.pcloud.link/publink/show?code=XZaxOLZWjuaFoeOEhbQ30Y9bFAvOyvQwol7

**Comment by BuzzwordChief** [comment]
Haha!

The build you send me running on the iGPU (i7 12700H, Intel Xe):
![grafik](https://user-images.githubusercontent.com/11695209/182439502-a222ac77-1bd5-4560-92cf-e998e14ae65c.png)

But when I force the build you send me to use the dGPU (Nvidia RTX 3050):
![grafik](https://user-images.githubusercontent.com/11695209/182439704-7bb72cc2-e11a-4aea-8f23-4d54ad525147.png)

So it seems it might be an Win11 / Intel Xe bug?

**Comment by BuzzwordChief** [comment]
I don't know if am stupid but now the "normal" 0.33.4 release from the website works if I force the Nvidia GPU. Intel Xe still has the transparent borders...

So maybe just an Intel Xe problem?

**Comment by slynch8** [comment]
Yes, I suspect something just needed to reset when you switched GPU. Doesn't sound like my change helped.

**Comment by slynch8** [comment]
It's difficult for me to fix without the right hardware. Here's a simple project with source that should show the problem in case you want to have a go yourself (apologies for the messy code)
[BorderTest.zip](https://github.com/slynch8/10x/files/9244541/BorderTest.zip)

**Comment by BuzzwordChief** [comment]
I looked through the code, tried some things, but I have to pass on this one. Not enough native windows development experience tbh. 

If I understand the example correctly it should just be a pitch black window with a shadow? I get a grayish window with the transparent border.

![grafik](https://user-images.githubusercontent.com/11695209/182448300-d33574f1-dc52-4cda-85db-c347e740dafa.png)

**Comment by slynch8** [comment]
ok, thanks for taking a look. It should be a grey window without any transparent border in case anyone else wants to try.

**Comment by slynch8** [comment]
potential fix, can someone try this build: https://t.co/l92cThvahy

**Comment by creikey** [comment]
> potential fix, can someone try this build: https://t.co/l92cThvahy

![image](https://user-images.githubusercontent.com/28424206/188754330-d01be5e1-c29a-456f-8d18-22960e870ea9.png)

Still has transparent borders

**Comment by slynch8** [comment]
oh well, thanks for trying

**Comment by slynch8** [comment]
ah, the fix will only work on Windows 11 Build 22621, which isn't on general release yet

**Comment by SirTimothy** [comment]
Hi, I tried BorderFix13 on my Win11 laptop (after updating to 22621.521 yesterday), still getting the transparent border.

Was playing around a bit with your BorderTest thing, it seems like setting `DWMWA_ALLOW_NCPAINT` is messing stuff up... that on its own is giving me a window with a very thick border and tall titlebar... it seems as though the non-client area is also getting extended into the client area (so having 8px non-client area from `WM_NCCALCSIZE`, end up with 8px border also leaking into the client area)... but that leak is limited by some amount, if I crank up the border size it still only leaks a little way in... for some reason handling `WM_NCPAINT` or doing `SetWindowRgn()` in `WM_ACTIVATE` makes the border go transparent...

Not sure if it's a bug with Win11, or maybe there's just something else missing there (I haven't done much with custom-drawn Win32)

If I skip the `DWMWA_ALLOW_NCPAINT`, then the main window area looks fine, but the drop shadow in the non-client area isn't transparent, so that's not so good... If it's just for a shadow, maybe it would work to just have a regular borderless window (with `WM_NCHITTEST of course to do drag/resize), and let the OS just do its own shadow?

**Comment by SirTimothy** [comment]
Actually, I just went back and tried what BuzzwordChief mentioned, forced 10X to run on GTX 1060 (instead of Intel UHD 630) and it seems to work properly. Multiple runs, switched setting a few times, consistently broken on iGPU, good on dGPU... So that seems like maybe an okay workaround, but seems overkill to push it to the higher power card just so DWM doesn't break everything (or whatever's happening there)...

**Comment by slynch8** [comment]
The problem with a regular borderless window is it doesn't support minimise/maximise and the task-bar operations. As you say if I remove DWMWA_ALLOW_NCPAINT then there is a black dropshadow. If I set the border size to 0 then I don't get any dropshadow and can't resize the window (although maybe I could add that myself). There must be some combination I'm missing where Windows can render the dropshadow, but I can render the titlebar area.

**Comment by slynch8** [comment]
Is anyone on Windows 11 Build 22621 yet?

**Comment by slynch8** [comment]
I'm trying a different approach, can you try this build
https://e1.pcloud.link/publink/show?code=XZrbs8ZUOSts8oabR03xGJmaSWJkS9K0Quy

For some reason it's lost the window resize at the top of the window but everything else seems to work fine. Maybe I can handle that myself.

**Comment by SirTimothy** [comment]
BorderFix14 seems to be working for me, without forcing any graphics settings! Windowed and maximized both seem okay. As you said, the top resize is missing (top corners are there though), but otherwise it seems to be pretty good :)

When maximized, the top menu feels pretty tight, like it's almost getting cut off. The text touches the very top of the screen, while windowed has some padding there. But I'm not sure if that's anything related to this, or maybe that's just how it is :)

<img width="305" alt="image" src="https://user-images.githubusercontent.com/6760500/193455941-711e7e40-9fab-42c9-a4a5-5aa4fcac3308.png">
<img width="788" alt="image" src="https://user-images.githubusercontent.com/6760500/193455981-f85a1bd5-274d-4c34-9805-5f300ea70205.png">
<img width="407" alt="image" src="https://user-images.githubusercontent.com/6760500/193456244-2602139b-449d-4f3d-a375-203cdbf690b9.png">

**Comment by slynch8** [comment]
That's good. Can you try this build. I've fixed the resize top and the maximised size

https://e1.pcloud.link/publink/show?code=XZlDs8Zn98VVXJAJMXoRnQbgd1hJyItrloX

**Comment by slynch8** [comment]
When not maximised should look like this, with a dropshadow and 1 pixel border

![image](https://user-images.githubusercontent.com/7141856/193458385-2456eb6a-ed00-42c1-8fdd-52757bf05701.png)

**Comment by slynch8** [comment]
and maximised it should be like this

![image](https://user-images.githubusercontent.com/7141856/193458476-be7d4877-c4ef-4c5f-bdeb-efff94e9b155.png)

**Comment by SirTimothy** [comment]
Windowed keeps the Win11 curved corners, and a thin outer border, but in general looks pretty good :) Also seems to behave reasonably with DPI scaling (I'm at 150%) :)

![image](https://user-images.githubusercontent.com/6760500/193458903-18bb9711-6de9-4f28-b12d-07bc52962bce.png)
![image](https://user-images.githubusercontent.com/6760500/193458919-8f6b3bf2-4838-4e1c-8816-8b25fb970f90.png)

There's one weird thing with maximized, I'm able to still use the top resize, which lets me adjust the window height but keeps a maximized state (according to the Maximize/Restore button in the top right), also has a thick white border outside the actual window, probably coming from Win11 theme. Very minor issue I'd say, it's awesome to have the clean borders :D

**Comment by slynch8** [comment]
Awesome! Shame about the curved borders but I can live with that :-) I've just fixed the resize top when maximised.

**Comment by slynch8** [comment]
It doesn't seem to have a dropshadow? but maybe that's normal on Windows 11

**Comment by slynch8** [comment]
Looks like it might be possible to remove the rounded corners if you want to test this in the test app
https://social.msdn.microsoft.com/Forums/office/en-US/c4c0c7e4-90a2-4037-85af-1e961dd20a62/remove-rounded-corner?forum=windowssdk

**Comment by SirTimothy** [comment]
> Shame about the curved borders but I can live with that :-)

Yeah no problem there, it's not cutting off anything. Curves aren't there when maximized or snapped (for some reason can't snap by dragging it to the edge of the screen, but Win+Left/Right works), so it basically looks consistent with anything else.

> It doesn't seem to have a dropshadow? but maybe that's normal on Windows 11

There's a shadow, but it's very subtle compared to other things (browser, explorer, etc). While resizing/moving the 10x window, I can see the brightness changing behind it, and with a screenshot I cranked contrast and saw that it is indeed there.

**Comment by slynch8** [comment]
Thanks for testing these builds. I'll get a new build uploaded soon.

**Comment by slynch8** [comment]
fixed (0.34.19)

**Comment by slynch8** [comment]
I need to make some modifications to this fix. Can someone check this exe for me, it should show a grey window with a green border
https://e1.pcloud.link/publink/show?code=XZ4Jn8ZXAGlw2rz7zkrwVYHowb3nVHFuR2X

![image](https://user-images.githubusercontent.com/7141856/193804242-9fb5040c-6cac-49d5-9abf-9d62365aabd0.png)

**Comment by BuzzwordChief** [comment]
No green border...

<img width="175" alt="image" src="https://user-images.githubusercontent.com/11695209/193837017-61e08910-6a3b-4df6-ab39-1ea5d0ff614f.png">

**Comment by slynch8** [comment]
That's probably not an issue, is it resizable?

**Comment by slynch8** [comment]
Great, thanks for checking

**Comment by slynch8** [comment]
This is hopefully the final build, would you give it a try please?
https://e1.pcloud.link/publink/show?code=XZB498Z4s7yisyOCkjAQ2E2cfdo4V5RzxbV

**Comment by BuzzwordChief** [comment]
No problem!

- In window mode the green border is drawn but only one or two pixels wide
- Windows key + arrow key works
- Fullscreen looks good
- Snapping the window by dragging it to edge of the screen does not work at all

Window Mode:
![image](https://user-images.githubusercontent.com/11695209/193944758-c269a946-0ad1-4c55-85dc-3d5a66a464c8.png)

Fullscreen Mode:
![image](https://user-images.githubusercontent.com/11695209/193944794-c5b11e3a-f860-4054-8eb5-b7f19b0a3f3a.png)

**Comment by BuzzwordChief** [comment]
Correction: Snapping the window fullscreen by dragging it to the top of the screen works BUT I get no visual indicator, e.g. this expanding animation. Also it only works on my external monitor!?

**Comment by slynch8** [comment]
Looks good! Thanks for checking. Yes, Win11 seems to trim off a pixel from the border, but I don't think there is anything I can do about that. I'll have a look at the drag-side docking next.

**Comment by slynch8** [comment]
fixed (0.34.20)

dragging to edges now works, along with the visual indicator
